org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/util/SoftClassCache.java

```diff
@@ -0,0 +1,85 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Christoph Läubrich.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Christoph Läubrich - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.jdt.internal.compiler.util;
+
+import java.io.IOException;
+import java.lang.ref.SoftReference;
+import java.nio.file.Path;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Implements a soft cache for reading class files from disk, as these caches can grow quite large but data can be
+ * recovered afterwards we only hold a soft reference to the bytes itself.
+ */
+class SoftClassCache {
+
+	private ConcurrentMap<String, JdkClasses> jdks = new ConcurrentHashMap<>();
+
+	void clear() {
+		this.jdks.clear();
+	}
+
+	public byte[] getClassBytes(Jdk jdk, Path path) throws IOException {
+		return this.jdks.computeIfAbsent(jdk.path, JdkClasses::new).get(path);
+	}
+
+	private static final class JdkClasses {
+		private ConcurrentMap<Path, ClassBytes> classes = new ConcurrentHashMap<>(10007);
+		private String jdkPath;
+
+		public JdkClasses(String jdkPath) {
+			this.jdkPath = jdkPath;
+		}
+
+		public byte[] get(Path path) throws IOException {
+			return this.classes.computeIfAbsent(path, ClassBytes::new).getBytes();
+		}
+
+		@Override
+		public String toString() {
+			return "Class Cache for " + this.jdkPath; //$NON-NLS-1$
+		}
+	}
+
+	private static final class ClassBytes {
+		private final Path path;
+		private boolean empty;
+		private SoftReference<byte[]> bytes;
+
+		public ClassBytes(Path path) {
+			this.path = path;
+		}
+
+		public synchronized byte[] getBytes() throws IOException {
```

Comment:
Please remove `synchronized` here, as the data we are supposed to read shouldn't change, we don't care if that is read only once or may be two times in parallel **initially** - for the cost of **permanent** locking on data retrieval. Note, the original code wasn't synchronized for same reason.

Reply:
The previous code *was* synchronized by the concurrent hash map therefore I found it might be a good idea here as well. if that's not desired I can change it but wanted to note it here just in case...

Reply:
> I can change it but wanted to note it here just in case...

Thanks, noted, please remove synchronized.

Reply:
Done.